pkg/store/bucket.go

```diff
@@ -824,11 +834,36 @@ func blockSeries(
 		}
 
 		s := seriesEntry{}
+		if err := indexr.LookupLabelsSymbols(symbolizedLset, &lset); err != nil {
+			return nil, nil, errors.Wrap(err, "Lookup labels symbols")
+		}
+		var tombstoneIntervals promtombstones.Intervals
+		for _, ts := range tombstones {
+			for _, matcher := range *ts.Matchers {
+				if val := lset.Get(matcher.Name); val != "" {
+					if !matcher.Matches(val) {
+						continue
+					}
+					if skipChunks {
+						continue PostingsLoop
+					}
+					tombstoneIntervals = tombstoneIntervals.Add(promtombstones.Interval{Mint: ts.MinTime, Maxt: ts.MaxTime})
+				}
+			}
+		}
```

Comment:
I am wondering which way is better. Now we are matching tombstones at the query time, which is online processing. Another way is to build an in-memory Prometheus tombstone for each matched block when we load tombstones each time. Since block contents are immutable, we can cache the in-mem tombstone results each time based on the tombstone ids.
Ideally, the second option is better in performance. But happy to hear more ideas.